mathjax3-ts/input/tex.ts

```diff
@@ -22,58 +22,323 @@
+    let rewrite: MmlNode[] = [];
+    node.walkTree((n, d) => {
+      const children = n.childNodes;
+      if ((n.isKind('msubsup') && (!children[(n as MmlMsubsup).sub] ||
+                                   !children[(n as MmlMsubsup).sup])) ||
+          (n.isKind('munderover') && (!children[(n as MmlMunderover).under] ||
+                                      !children[(n as MmlMunderover).over]))) {
+        d.unshift(n);
+      }
+    }, rewrite);
+    for (const n of rewrite) {
+      const children = n.childNodes as (MmlNode|TextNode)[];
+      const parent = n.parent;
+      let ms, newNode;
+      if (n.isKind('msubsup')) {
+        ms = n as MmlMsubsup;
+        newNode = (children[ms.sub] ?
+                   options.nodeFactory.create('node', 'msub', [children[ms.base], children[ms.sub]], {}) :
+                   options.nodeFactory.create('node', 'msup', [children[ms.base], children[ms.sup]], {}));
+      } else {
+        ms = n as MmlMunderover;
+        newNode = (children[ms.under] ?
+                   options.nodeFactory.create('node', 'munder', [children[ms.base], children[ms.under]], {}) :
+                   options.nodeFactory.create('node', 'mover', [children[ms.base], children[ms.over]], {}));
+      }
+      NodeUtil.copyAttributes(n, newNode);
+      // This is only necessary if applied to an incomplete node, where
+      // msubsup/underover can be top level.
+      if (parent) {
+        parent.replaceChild(newNode, n);
+      } else {
+        node = newNode;
+      }
     }
+  };
+
 
-    /*
-     * @override
-     */
-    public findMath(strings: string[]) {
-        return this.findTeX.findMath(strings);
+  /**
+   * @override
+   */
+  constructor(options: OptionList) {
+    let [tex, find] = separateOptions(options, FindTeX.OPTIONS);
+    super(tex);
+    this.findTeX = this.options['FindTeX'] || new FindTeX(find);
+  }
+
+  /**
+   * @override
+   */
+  public compile(math: MathItem<N, T, D>): MmlNode {
+    this.parseOptions = this.configure();
+    let node: MmlNode;
+    let parser: TexParser;
+    let display = math.display;
+    this.latex = math.math;
+    this.runPreprocessors();
+    try {
+      parser = new TexParser(this.latex,
+                             {display: display, isInner: false},
+                             this.parseOptions);
+      node = parser.mml();
+    } catch (err) {
+      if (!(err instanceof TexError)) {
+        throw err;
+      }
+      node = this.formatError(err);
+    }
+    this.mathNode = this.parseOptions.createNode('node', 'math', [node], {});
+    if (display) {
+      NodeUtil.setAttribute(this.mathNode, 'display', 'block');
     }
+    TeX.cleanSubSup(this.mathNode, this.parseOptions);
+    this.mathNode.setInheritedAttributes({}, display, 0, false);
+    TeX.cleanStretchy(this.mathNode, this.parseOptions);
+    this.mathNode.setInheritedAttributes({}, display, 0, false);
+    this.mathNode.setTeXclass(null);
```

Comment:
Thanks for the explanation. I thought since it was run in `tex-json.js` it should be a run as a finalisation. I removed it now but run it in `parse-tests.js`.